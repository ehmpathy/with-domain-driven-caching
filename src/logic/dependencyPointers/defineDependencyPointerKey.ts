import { sha256 } from 'cross-sha256';
import {
  DomainEntity,
  DomainValueObject,
  getUniqueIdentifierSlug,
} from 'domain-objects';
import { PickOne } from 'type-fns';
import { SerializableObject } from 'with-cache-normalization/dist/domain/NormalizeCacheValueMethod';

const serializePropertyValue = (value: SerializableObject) => {
  // if it is a domain object, get its unique identifier
  const isDomainIdentifiable =
    value instanceof DomainEntity || value instanceof DomainValueObject;
  if (isDomainIdentifiable) return getUniqueIdentifierSlug(value);

  // otherwise, must use JSON.stringify, to avoid getting things like `[ Object: object ]`; however, must be filePathSafe, so cross-sha hash and include human part
  const humanPart = JSON.stringify(value)
    .replace(/:/g, '.')
    .replace(/[^\w\-\_]/g, '')
    .replace(/\.\./g, '.');
  const uniquePart = new sha256().update(JSON.stringify(value)).digest('hex'); // part to guarantee uniqueness
  return [humanPart, uniquePart].join('.');
};

/**
 * defines a dependency pointer key
 *
 * supports both cases
 * - foreignKey specified on source table
 * - foreignKey specified on target table
 */
export const defineDependencyPointerKey = ({
  dobj,
  property,
  specifier,
}: {
  dobj: string;
  property: string;
  specifier: PickOne<{
    /**
     * specifies that the dependency is on *any* value of the property from this *specific* domain object
     */
    propertyOf: { uuid: string };

    /**
     * specifies that the dependency is on this *specific* value of the property from *any* domain object of this class
     */
    propertyEquals: { value: SerializableObject };
  }>;
}): string =>
  `.query.dep.${[
    dobj,
    specifier.propertyOf ? `.uuid.${specifier.propertyOf.uuid}` : '',
    '.',
    property,
    specifier.propertyEquals
      ? `.${serializePropertyValue(specifier.propertyEquals.value)}`
      : '',
  ].join('')}`;
