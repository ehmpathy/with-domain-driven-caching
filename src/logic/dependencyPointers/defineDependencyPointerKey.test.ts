import { DomainEntity, DomainLiteral } from 'domain-objects';

import { defineDependencyPointerKey } from './defineDependencyPointerKey';

interface ContainerLock {
  manufacturer: string;
  model: string;
}
class ContainerLock
  extends DomainLiteral<ContainerLock>
  implements ContainerLock {}

interface Container {
  icid: string;
  onShipUuid: string;
  lock: ContainerLock;
}
class Container extends DomainEntity<Container> implements Container {}

describe('defineDependencyPointerKey', () => {
  it('should be able to define the dependency pointer key for a relationship from entity with foreign key on *source* entity', () => {
    // specify a relationship from container.uuid=x to ship.uuid=any
    const key = defineDependencyPointerKey({
      dobj: Container.name,
      property: 'onShipUuid',
      specifier: {
        propertyOf: { uuid: '__uuid__' },
      },
    });
    expect(key).toEqual('.query.dep.Container.uuid.__uuid__.onShipUuid');
    expect(key).toMatchSnapshot();
  });
  it('should be able to define the dependency pointer key for a relationship from entity with foreign key on *target* entity', () => {
    // specify a relationship from container.uuid=x to ship.uuid=any
    const key = defineDependencyPointerKey({
      dobj: Container.name,
      property: 'onShipUuid',
      specifier: {
        propertyEquals: { value: '__uuid__' },
      },
    });
    expect(key).toEqual(
      '.query.dep.Container.onShipUuid.__uuid__.c72319fa41f2824995e59ec2f82e14eda8a5a017cdc6f9b971a02e205d306813',
    );
    expect(key).toMatchSnapshot();
  });
  it('should be able to define the dependency pointer key to a domain value object property', () => {
    const lock = new ContainerLock({
      manufacturer: 'swiper no swiping',
      model: 'locky mclockface',
    });

    // specify a relationship from container.uuid=x to ship.uuid=any
    const key = defineDependencyPointerKey({
      dobj: Container.name,
      property: 'lock',
      specifier: {
        propertyEquals: { value: lock },
      },
    });
    expect(key).toEqual(
      '.query.dep.Container.lock.ContainerLock.swipernoswiping.lockymclockface.3cf448f9380ecd4daf668110f1f0bf7884de32b69c3576082bf0dcc1ca50c2b9',
    );
    expect(key).toMatchSnapshot();
  });
});
