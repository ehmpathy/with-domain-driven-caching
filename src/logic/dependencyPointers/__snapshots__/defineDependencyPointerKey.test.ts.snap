// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineDependencyPointerKey should be able to define the dependency pointer key for a relationship from entity with foreign key on *source* entity 1`] = `".query.dep.Container.uuid.__uuid__.onShipUuid"`;

exports[`defineDependencyPointerKey should be able to define the dependency pointer key for a relationship from entity with foreign key on *target* entity 1`] = `".query.dep.Container.onShipUuid.__uuid__.c72319fa41f2824995e59ec2f82e14eda8a5a017cdc6f9b971a02e205d306813"`;

exports[`defineDependencyPointerKey should be able to define the dependency pointer key to a domain value object property 1`] = `".query.dep.Container.lock.ContainerLock.swipernoswiping.lockymclockface.3cf448f9380ecd4daf668110f1f0bf7884de32b69c3576082bf0dcc1ca50c2b9"`;
